api/include/opentelemetry/logs/logger.h

```diff
@@ -246,10 +247,225 @@ class Logger
     this->EmitLogRecord(Severity::kFatal, std::forward<ArgumentType>(args)...);
   }
 
+  //
+  // OpenTelemetry C++ user-facing Logs API
+  //
+
+  inline bool Enabled(Severity severity, const EventId &event_id) const noexcept
+  {
+    OPENTELEMETRY_LIKELY_IF(Enabled(severity) == false) { return false; }
+    return EnabledImplementation(severity, event_id);
+  }
+
+  inline bool Enabled(Severity severity, int64_t event_id) const noexcept
+  {
+    OPENTELEMETRY_LIKELY_IF(Enabled(severity) == false) { return false; }
+    return EnabledImplementation(severity, event_id);
+  }
+
+  inline bool Enabled(Severity severity) const noexcept
+  {
+    return static_cast<uint8_t>(severity) >= OPENTELEMETRY_ATOMIC_READ_8(&minimum_severity_);
+  }
+
+  /**
+   * Log an event
+   *
+   * @severity severity of the log
+   * @event_id event identifier of the log
+   * @format an utf-8 string following https://messagetemplates.org/
+   * @attributes key value pairs of the log
+   */
+  virtual void Log(Severity severity,
+                   const EventId &event_id,
+                   nostd::string_view format,
+                   const common::KeyValueIterable &attributes) noexcept
+  {
+    this->EmitLogRecord(severity, event_id, format, attributes);
+  }
+
+  virtual void Log(Severity severity,
+                   int64_t event_id,
+                   nostd::string_view format,
+                   const common::KeyValueIterable &attributes) noexcept
+  {
+    this->EmitLogRecord(severity, event_id, format, attributes);
+  }
+
+  virtual void Log(Severity severity,
+                   nostd::string_view format,
+                   const common::KeyValueIterable &attributes) noexcept
+  {
+    this->EmitLogRecord(severity, format, attributes);
+  }
+
+  virtual void Log(Severity severity, nostd::string_view message) noexcept
+  {
+    this->EmitLogRecord(severity, message);
+  }
+
+  // Convenient wrappers based on virtual Log() virtual methods.
+  // https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md#field-severitynumber
+
+  inline void Trace(const EventId &event_id,
+                    nostd::string_view format,
+                    const common::KeyValueIterable &attributes) noexcept
+  {
+    this->Log(Severity::kTrace, event_id, format, attributes);
+  }
```

Comment:
I am confused about the intended usage for this helper:

Is the caller supposed to check Enabled for each call, as in:

```
if (logger->Enabled(xxx)) {
  logger->Trace(xxx);
}
```

or is Trace() responsible to check the severity flag itself, as in:

```
logger->Trace(xxx);
```

The benchmarks shows both usage, is this code writing un wanted traces in the second case ?
The check for Enabled() seems missing from the code path.

Beside, if the whole point of Logger::minimum_severity_ is to check the flags in the API before making a virtual function call, then why making calls here to virtual functions Log() and EmitLogRecord() ?

Suggested changes:

* Call Enabled() in every user facing api.
* Do not call a user facing api like Log() from another user facing api like Trace(), call EmitLogRecord directly, so the Enabled() flags are not checked multiple times.

For example:

```
inline void Trace(xxx) {
  if (Enabled(xxx)) {
    EmitLogRecord(xxx);
  }
}
```

Reply:
Thanks for the suggestion @marcalff . 

> Call Enabled() in every user facing api.
`Enabled()` should be called by the user's code instead of from our all our wrappers which is intentional. The idea is to avoid calculating any parameters which is only for logging. In the simple case, `common::MakeAttributes({"key1", “value1}, {"key2", "value2"}}` takes significant extra time than the `Enabled()` check. And there could be many other heavy query result which passed as parameters to query. Calling `Enabled()` in user's code can eliminate such heavy query from the first place.

```c++
logger->Trace("some message template", common::MakeAttributes({"some_value_takes_time_calculate", long_query()}));
``` 

If we are not going to call `Enabled()` in user facing API, then inline the call of `DoLogRecord` seems unnecessary for all the wrappers then? I don't have strong preference as the name implied they are just wrappers. But I can also revisit this part in the SDK support because changing the implementation will not break the user facing API.

Please let me know if anything is missed here.

Reply:
Just want to have a discuss.
We have a log macro like this before in our project.

```cpp
#define WLOGDEBUG(logger, ...) \
  if(logger.check_level(log_level::DEBUG)) {
    logger->log(log_level::DEBUG, __VA_ARGS__);
  }
```

And when we use `WLOGDEBUG(logger, "Debug message: {}", protobuf_message.DebugString());` and if the debug log is not enabled, `protobuf_message.DebugString()` will not be called and it's a high cost function.

Is there any way we can achieve a similar effect?

Reply:
Perhaps we can define similar macro to help the user reduce boilerplate code, but it can avoid the logging cost. Based on the logging statement you shared, the the variable `protobuf_message` might be only need for logging, and it could do heavy initialization (like map or load string resources). This happens in the definition of `protobuf_message` which cannot be be optimized here.

```c++
  WLOGDEBUG(logger, "Debug message: {}", protobuf_message.DebugString());
```

While providing macros could help, the pointer in our API is make `Enabled` check explicit to the user, so they are aware of this and encouraged to do the optimization to archive high performance logging.

Reply:
Let's continue the discussion about Enabled() in the next patch, for the SDK part.

The fact that the caller MAY use Enabled() to optimize paths that evaluate complex arguments is good.

Still we can't force that the caller MUST use Enabled() all the time, so the implementation will need to check it again at some point.